chapters/VK_EXT_present_timing/queries.adoc

```diff
@@ -0,0 +1,481 @@
+were returned.
+
+include::{generated}/validity/protos/vkGetSwapchainTimeDomainsEXT.adoc[]
+--
+
+[open,refpage='VkSwapchainTimeDomainPropertiesEXT',desc='An available time domain for a swapchain',type='structs']
+--
+
+The sname:VkSwapchainTimeDomainPropertiesEXT structure is defined as:
+
+include::{generated}/api/structs/VkSwapchainTimeDomainPropertiesEXT.adoc[]
+
+  * pname:sType is the type of this structure.
+  * pname:pNext is `NULL` or a pointer to a structure extending this
+    structure.
+  * pname:timeDomain is a elink:VkTimeDomainEXT value representing a time
+    domain that is available for the swapchain.
+
+include::{generated}/validity/structs/VkSwapchainTimeDomainPropertiesEXT.adoc[]
+--
+
+[open,refpage='vkSetSwapchainTimingEXT',desc='Set timing information for a swapchain',type='protos']
+--
+
+To set timing information for a swapchain, call:
+
+include::{generated}/api/protos/vkSetSwapchainTimingEXT.adoc[]
+
+  * pname:device is the device associated with pname:swapchain.
+  * pname:swapchain is the swapchain to obtain timing properties for.
+  * pname:pSwapchainTimingInfo is `NULL` or a pointer to an instance of
+    the sname:VkSwapchainTimingInfoEXT structure.  If `NULL`,
+    display timing is disabled for the swapchain, otherwise enables
+    display timing and specifies which time domain to use.
+
+include::{generated}/validity/protos/vkSetSwapchainTimingEXT.adoc[]
+--
+
+[open,refpage='VkSwapchainTimingInfoEXT',desc='Specify which of the available time domains to use for a swapchain',type='structs']
+--
+
+The sname:VkSwapchainTimingInfoEXT structure is defined as:
+
+include::{generated}/api/structs/VkSwapchainTimingInfoEXT.adoc[]
+
+  * pname:sType is the type of this structure.
+  * pname:pNext is `NULL` or a pointer to a structure extending this
+    structure.
+  * pname:timeDomain is a elink:VkTimeDomainEXT value representing the time
+    domain that should be used with the swapchain.
+  * pname:presentAtNearestRefreshCycle specifies whether the presentation is
+    allowed to target the refresh cycle that is nearest to the target present
+    time, when one is provided. If ename:VK_FALSE, a targeted present stage
+    must: not complete before the target present time specified via
+    slink:VkPresentTimeEXT::pname:targetPresentTime.
+
+include::{generated}/validity/structs/VkSwapchainTimingInfoEXT.adoc[]
+--
+
+[open,refpage='vkGetPastPresentationTimingEXT',desc='Obtain timing of a previously-presented image',type='protos']
+--
+
+The implementation maintains a limited amount of history of timing
+information about previous presents, specified by
+slink:VkSwapchainPresentTimingCreateInfoEXT::pname:presentTimingQueueSize
+when creating a swapchain.
+Because of the asynchronous nature of the presentation engine, the timing
+information for a given flink:vkQueuePresentKHR command only becomes
+available some time later.
+These time values should: be asynchronously queried, and are returned if
+available.
```

Comment:
This is a really good point. In wayland VK_KHR_present_wait actually waits for the feedback to become available (because that's also the information the compositor has) but that's not defined to be exactly when the presentation actually happened but some time after it.

Reply:
Or before it, there is nothing stopping predicted presentation timestamps fwiu.

Right now Gamescope would actually give you it at latch time when we go to prepare everything with a predicted vblank time. (Same for our google display timing, you can get past presentation feedback *before* you actually get presented but that actually works out really well for clients in our experience)

Reply:
How strongly can Gamescope guarantee those predicted vblanks are hit though? That seems to violate the intent of the spec if it's not a 100% guarantee.

Reply:
It should be possible to predict the latch timestamp as soon as the QueuePresent semaphore is signaled. Later, composition can still be unexpectedly delayed, but that only affects FIRST_PIXEL_OUT and not the latch timestamp.

Reply:
Strong enough that we feel doing this prediction is 100% reliable and consistent in the steady 99.9% state.

The only edgecases there are in user interactive system states eg. where we might enable a blur for bringing up the Steam overlay, or the user toggles FSR on for a first frame before the dynamic vblank timing stuff pushes back the next frame.

From our perspective in Gamescope/SteamOS, it is likely best that the application does not see such a thing and start to adjust it's timings as we will have already fixed it by the next vblank behind their back.

(This applies for both present wait and display timing)

The implementation has been tested so far in Counter Strike 2 and Dota 2 and has been working really well, but we would definitely like more usecases to make sure we are doing the right thing that works best for application.

Reply:
This can only work with FRR. Predicting VRR is basically impossible in our user space right now.

Reply:
Yeah, I noticed that too. I didnt write the support for VRR there yet, but that one won't be predicted (and also needs to get stuff waking up based on like a timerfd or sooomething)

Reply:
@cubanismo Good points. I don't have a great answer to those questions. I did add an interaction with `vkWaitForPresentKHR` at some point, but I changed my mind after thinking about it some more: `vkWaitForPresentKHR` can return earlier than the present timing being available (predictions notwithstanding, which I hadn't considered before). Like you suggested, I think the present stats collection should be independent from the presentation itself.

> how do applications pick a reasonable queue size?
I don't see a good way to determine this, other than picking some default value and presenting a few times. In the end, it requires some kind of discovery of the presentation topology, and I see no other way to do this without getting into an unreasonable rabbit hole.

> It seems like we need some sort of forward progress guarantee at least. (...) What's to prevent an implementation from never making results available and claiming conformance?

That's probably the strongest argument for having some form of blocking call. What if `vkGetPastPresentationTimingEXT` blocked for elements of `pPresentationTimings` where `presentId` is non-0? As a side-effect, this would require `vkGetPastPresentationTimingEXT` not to be externally synchronized anymore, I think, which has been requested a couple times as well.